### PR TITLE
Generate loops based on the IdModel Loop graph (promotion)

### PR DESCRIPTION
**Pull request by naoyam**
This PR introduces an option to use loop promotion IDs when generating `ForLoop` IR nodes. Consequently, loop index variables are also maintained by `IdModel`, not by `ComputeAtMap`. 

This is required when loop domains are not directly derived from logical domains as the ComputeAtMap loop map does not support such loop domains.

For now, I decided to add yet another IdModel option (`IdModelEnableOption::Loop`) for more controlled, incremental progress. It is automatically enabled in most of the tests in `test_indexing_advanced.cpp`.

Major changes are:

- Concrete loop domain: it's either the promotion ID (when `IdModelEnableOption::Loop` is on) or the concrete ID in `ComputeAtMap`
- Loop index variable: it's either the variable managed by `IdModel` (when `IdModelEnableOption::Loop` is on) or `ComputeAtMap`

This is part of #2851.

### CODE REVIEW

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build --diff

**Comment by naoyam** [comment]
!build --diff

**Comment by naoyam** [comment]
!build --diff

**Comment by naoyam** [comment]
!build --diff

**Comment by naoyam** [comment]
!build --diff

**Comment by naoyam** [comment]
!build --diff

File: csrc/id_model/indexing_utils.h [hunk 1]
```diff
@@ -36,25 +36,6 @@ inline ForLoop* getForLoop(
   }
 }
 
-// Get the promotion domain of a given loop domain.
-inline IterDomain* getLoopPromotion(
```
Comment (naoyam) [inline csrc/id_model/indexing_utils.h:40]: Just moved to id_model/utils.h

File: csrc/id_model/indexing.cpp [hunk 2]
```diff
@@ -791,56 +769,17 @@ void TensorIndexer::buildLoopIndexMap() {
...
-        }
+        loop_index = GpuLower::current()->getLoopIndexVariable(loop_id);
       }
 
       loop_index_map_[loop_group] = loop_index;
     }
   }
 }
 
-bool TensorIndexer::shouldUseZeroIndex(const ValGroup& loop_group) const {
```
Comment (naoyam) [inline csrc/id_model/indexing.cpp:832]: Moved to id_model/utils.h

File: csrc/id_model/indexing.cpp [hunk 3]
```diff
@@ -731,28 +731,6 @@ class AllocationDomainSetup : private kir::IrVisitor {
   std::unordered_set<TensorView*> used_as_producer;
 };
 
-ParallelType getParallelType(const ValGroup& loop_group) {
```
Comment (naoyam) [inline csrc/id_model/indexing.cpp:734]: Moved to id_model/utils.h

File: csrc/id_model/utils.h [hunk 4]
```diff
@@ -59,4 +85,62 @@ inline bool isIdModelOptionEnabled(IdModelEnableOption option) {
   return opts.find(option) != opts.end();
 }
 
+// Get the promotion domain of a given loop domain.
```
Comment (naoyam) [inline csrc/id_model/utils.h:88]: Just move here

File: csrc/id_model/id_model.cpp [hunk 5]
```diff
@@ -1007,4 +1012,123 @@ void IdModel::validateAndPropagatePType() {
...
+      //  lowering phase so this function will be querried before the double
+      //  buffer pass. At that point, no forloop has any circular buffer
+      //  stage defined, and we just default to using the main stage index.
+      circular_buffer_loop_stage = CircularBufferLoopStage::Main;
+    }
+    return circular_buffered_loop_index_variable_map_.at(loop_group)
+        ->at(circular_buffer_loop_stage);
+  } else {
+    return loop_index_variable_map_.at(loop_group);
+  }
```
Comment (naoyam) [inline csrc/id_model/id_model.cpp:1124]: Just follows what the current ComputeAtMap does.

File: csrc/id_model/indexing.cpp [hunk 6]
```diff
@@ -791,56 +769,17 @@ void TensorIndexer::buildLoopIndexMap() {
...
-          // should be used, but that means we would need to consider
-          // multiple outputs with different memory types, though it
-          // should be uncommon in practice.
-          shouldUseZeroIndex(loop_group) || isParallelTypeDeviceDim(ptype)) {
+      if (shouldUseZeroIndex(loop_group, id_model_)) {
         loop_index = fusion->zeroVal();
-      } else if (isParallelTypeThread(ptype)) {
-        loop_index = NamedScalar::getParallelIndex(ptype);
       } else {
-        // Until the transition to the IdModel-based indexing is
```
Comment (naoyam) [inline csrc/id_model/indexing.cpp:805]: This logic is moved to `IdModel::allocateLoopIndexVariables`

File: csrc/device_lower/utils.cpp [hunk 7]
```diff
@@ -783,8 +785,8 @@ BasicAllocInfo getAllocInformation(
       }
     }
 
-    if (GpuLower::current()->caMap()->areMapped(
-            local_id, fl_id, IdMappingMode::PERMISSIVE)) {
+    if (lower_utils::getConcreteLoopID(local_id) ==
```
Comment (naoyam) [inline csrc/device_lower/utils.cpp:788]: This part needs to be changed as we are no longer using the CA map if the option is enabled.

**Review by zasdfgbnm — APPROVED** [review]

**Comment by naoyam** [comment]
!build

**Pull request merged**